lib/spack/spack/test/zz/bindist_rel.py

```diff
@@ -0,0 +1,379 @@
+    real_store = spack.store.store
+    real_layout = spack.store.layout
+    spack.store.store = spack.store.Store(str(tmpdir.join('opt')))
+    spack.store.layout = YamlDirectoryLayout(str(tmpdir.join('opt')),
+                                             path_scheme=def_install_path_scheme)  # noqa: E501
+    yield spack.store
+    tmpdir.join('opt').remove()
+    spack.store.store = real_store
+    spack.store.layout = real_layout
+
+
+@pytest.fixture(scope='function')
+def install_dir_non_default_layout(tmpdir):
+    """Hooks a fake install directory with a non-default directory layout"""
+    real_store = spack.store.store
+    real_layout = spack.store.layout
+    spack.store.store = spack.store.Store(str(tmpdir.join('opt')))
+    spack.store.layout = YamlDirectoryLayout(str(tmpdir.join('opt')),
+                                             path_scheme=ndef_install_path_scheme)  # noqa: E501
+    yield spack.store
+    tmpdir.join('opt').remove()
+    spack.store.store = real_store
+    spack.store.layout = real_layout
+
+
+@pytest.mark.maybeslow
+@pytest.mark.usefixtures('gpg_directory', 'default_config',
+                         'install_dir_default_layout', 'cache_directory')
+def test_buildcache_create_install_rel(tmpdir, mirror_directory_rel,
+                                       mock_stage):
+    """
+    Test the creation and installation of buildcaches with relative
+    rpaths into the default directory layout scheme.
+    """
+    global mirror_path_rel
+    mirror_path_rel = mirror_directory_rel
+    mparser = argparse.ArgumentParser()
+    mirror.setup_parser(mparser)
+    margs = mparser.parse_args(
+        ['add', '--scope', 'site', 'test-mirror-rel', 'file://%s' % mirror_path_rel])
+    mirror.mirror(mparser, margs)
+    margs = mparser.parse_args(['list'])
+    mirror.mirror(mparser, margs)
+    cparser = argparse.ArgumentParser()
+    compiler.setup_parser(cparser)
+    cargs = cparser.parse_args(['add', '--scope', 'site'])
+    compiler.compiler(cparser, cargs)
+    cargs = cparser.parse_args(['list', '--scope', 'site'])
+    compiler.compiler(cparser, cargs)
+    cargs = cparser.parse_args(['info', 'clang'])
+    compiler.compiler(cparser, cargs)
+    cargs = cparser.parse_args(['info', 'gcc'])
+    compiler.compiler(cparser, cargs)
+
+    rparser = argparse.ArgumentParser()
+    repocmd.setup_parser(rparser)
+    rargs = rparser.parse_args(['add', spack.paths.packages_path])
+    repocmd.repo(rparser, rargs)
+    njobs = spack.config.get('config:build_jobs')
+    if not njobs:
+        spack.config.set('config:build_jobs', 4, scope='user')
+    extensions = spack.config.get('config:template_dirs')
+    if not extensions:
+        spack.config.set('config:template_dirs',
+                         [os.path.join(spack.paths.share_path, 'templates')],
+                         scope='user')
+    espec = Spec('environment-modules')
+    espec.concretize()
+    zspec = Spec('zlib')
```

Comment:
Would it be possible to create a buildcache for a mock package? I think that would go faster (e.g. no setting of build jobs required) and be less fragile.

Reply:
Its better to work with a real example like environment modules which depends on tk. Tk is the one with the config path compiled in and is a very good test of bin string replacement.

Reply:
I was suggesting generating a binary file which is equivalent to the one produced by `tk`, at least for the purposes of the test: I figured you could write some number of binary 0's, followed by a path, and then some more into a file and this would test the `tk` instance. Perhaps that doesn't capture the complexity (e.g. a valid elf header is required), in which case I think we could consider pre-generating that exe with a small example and distributing that binary file as part of the Spack tests (so we don't actually have to Spack install anything).

In short: I agree that this properly tests the complexity and that is important, but if we can achieve the same effect without installing things, the tests will run faster, and I also think it will be clearer what they are testing.

Reply:
Well, OK. I initially had added buildcache binaries to the pull request.

Reply:
Now I recall the context. I wanted to add a macOS buildcache to test installing on linux with macholib.